### PR TITLE
Publisher system for kernel events in the kernel extension.

### PR DESCRIPTION
**Pull request by mofarrell**
Created a more generic way to add kernel events.

### CODE REVIEW

File: kernel/src/osquery.cpp [hunk 1]
```diff
@@ -365,6 +301,12 @@ static int osquery_ioctl(dev_t dev, u_long cmd, caddr_t data,
     case OSQUERY_IOCTL_BUF_ALLOCATE:
       alloc = (osquery_buf_allocate_args_t *)data;
 
+      if (alloc->version != OSQUERY_KERNEL_COMM_VERSION) {
```
Comment (theopolis) [inline kernel/src/osquery.cpp]: Woot!

File: kernel/src/publishers.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <feeds.h>
+#include "circular_queue_kern.h"
+
+typedef int (*osquery_subscriber_t)(osquery_cqueue_t *queue, void *udata);
```
Comment (theopolis) [inline kernel/src/publishers.h]: Docblocks?

File: kernel/src/osquery.cpp [hunk 3]
```diff
@@ -342,7 +278,7 @@ static int osquery_ioctl(dev_t dev, u_long cmd, caddr_t data,
   switch (cmd) {
     case OSQUERY_IOCTL_SUBSCRIPTION:
       sub = (osquery_subscription_args_t *)data;
-      if ((err = subscribe_to_event(sub->event, sub->subscribe))) {
+      if ((err = subscribe_to_event(sub->event, sub->subscribe, sub->udata))) {
```
Comment (theopolis) [inline kernel/src/osquery.cpp]: Is there any way to guarantee a size of `data`? Or at least to make sure it's not `== NULL`?

Comment (mofarrell) [inline kernel/src/osquery.cpp]: Apple IOCTL assure this.  udata will only be supplied by particular subscription types, and will need to be checked and copied into kernel space by the kernel land publisher.

**Pull request merged**